api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -447,15 +438,77 @@ public static NestedField of(int id, boolean isOptional, String name, Type type,
       return new NestedField(isOptional, id, name, type, doc, null, null);
     }
 
-    public static NestedField of(
-        int id,
-        boolean isOptional,
-        String name,
-        Type type,
-        String doc,
-        Object initialDefault,
-        Object writeDefault) {
-      return new NestedField(isOptional, id, name, type, doc, initialDefault, writeDefault);
+    public static Builder from(NestedField field) {
+      return new Builder(field);
+    }
+
+    public static Builder required(String name) {
+      return new Builder(false, name);
+    }
+
+    public static Builder optional(String name) {
+      return new Builder(true, name);
+    }
+
+    public static class Builder {
+      private final boolean isOptional;
+      private final String name;
+      private Integer id = null;
+      private Type type = null;
+      private String doc = null;
+      private Object initialDefault = null;
+      private Object writeDefault = null;
+
+      private Builder(boolean isOptional, String name) {
+        this.isOptional = isOptional;
+        this.name = name;
+      }
+
+      private Builder(NestedField toCopy) {
+        this.isOptional = toCopy.isOptional;
+        this.name = toCopy.name;
+        this.id = toCopy.id;
+        this.type = toCopy.type;
+        this.doc = toCopy.doc;
+        this.initialDefault = toCopy.initialDefault;
+        this.writeDefault = toCopy.writeDefault;
+      }
+
+      public Builder withId(int id) {
+        this.id = id;
+        return this;
+      }
+
+      public Builder ofType(Type type) {
+        this.type = type;
+        return this;
+      }
+
+      public Builder withDoc(String doc) {
+        this.doc = doc;
+        return this;
+      }
+
+      public Builder withInitialDefault(Object initialDefault) {
+        this.initialDefault = initialDefault;
+        return this;
+      }
+
+      public Builder withWriteDefault(Object writeDefault) {
+        this.writeDefault = writeDefault;
+        return this;
+      }
+
+      public Builder withDefault(Object defaultValue) {
+        this.initialDefault = defaultValue;
+        this.writeDefault = defaultValue;
+        return this;
+      }
```

Comment:
I would skip this method since semantics of initial and write default are too different. Further, initial cannot change but write can. Also, we may get unexpected results of chaining this with the other two methods.

Reply:
I thought about these issues originally, but ended up not adding validations to mirror the behavior of the original version that used factory methods. I think it's fine to have a method that sets both as a convenience.

The question is whether to enforce default value behavior here or elsewhere. I think the validation should be done elsewhere because:
1. As I noted above, the builder is doing the same thing as the factory method -- there was no validation before
2. There isn't enough information to validate here. This _could_ only allow setting the `initialDefault` once, but what does that accomplish? It only limits the builder API by assuming what the caller is doing when constructing a field. It would be perfectly valid to use this builder to copy another field (which sets `initialDefault`) and then change the initial default to create a new field with a different one. Schema validation is not the builder's responsibility.